WordPress/gutenberg
# Refactored Gallery: Add custom gutter size feature

## Description
This allows for the user to specify a custom gutter size for the new refactored gallery block.

## Background
The approach for this feature was originally implemented as part of the gallery refactoring. It has been split out so it can be discussed independently. A few blocks would benefit from being able to implement custom gutter sizes so a unified approach would be best, whether this is via the block supports mechanism or something more specific.

A non-gallery example of the need for custom gutter sizes can be found in the Columns block. There is already an [outstanding PR adding custom gutter sizing](https://github.com/WordPress/gutenberg/pull/27661) for it, using a similar ad hoc approach to this PR.

The [discussion related to this feature](https://github.com/WordPress/gutenberg/pull/25940#discussion_r574299055) on the refactored gallery PR can be found here: https://github.com/WordPress/gutenberg/pull/25940#discussion_r574299055

## Possible Options

- Possibly leverage the block supports mechanism to provide the gutter margin values if the next evolution of block supports allows specific targeting of inner elements or better yet to apply the value as a single CSS variable/custom property.
- A dedicated new hook that can offer the required gutter size settings etc in a consistent way regardless of which block is using it.
- Ad hoc approaches that fill each block's specific needs directly.

## How has this been tested?
Manually.

#### Testing Instructions
1. Checkout this PR branch and ensure the refactored gallery Gutenberg experiment is turned on.
2. Create a new post and add a gallery with images
3. Select the gallery and adjust the gutter size control in the sidebar, confirming the gallery visually updates as expected
4. Save the post and confirm the gutter sizing is correct in the frontend

## Screenshots
![GalleryGutters](https://user-images.githubusercontent.com/60436221/108705730-6c08fe80-7559-11eb-914a-ab3b8fef2fcc.gif)

## Types of changes
New feature

## Checklist:
